StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1117ParametersMustBeOnSameLineOrSeparateLines.cs

```diff
@@ -55,24 +59,319 @@ public class SA1117ParametersMustBeOnSameLineOrSeparateLines : DiagnosticAnalyze
+            }
+        }
+
+        private static void Analyze(SyntaxNodeAnalysisContext context, SeparatedSyntaxList<ParameterSyntax> parameters)
+        {
+            ParameterSyntax previousParameter = parameters[1];
+            int firstParameterLine = parameters[0].GetLine();
+            int previousLine = previousParameter.GetLine();
+            Func<int, int, bool> lineCondition;
+
+            if (firstParameterLine == previousLine)
+            {
+                // parameters must be on same line
+                lineCondition = (param1Line, param2Line) => param1Line != param2Line;
+            }
+            else if (previousLine - firstParameterLine == 1)
+            {
+                // each parameter must be on its own line
+                lineCondition = (param1Line, param2Line) => param1Line == param2Line;
+            }
+            else
+            {
+                // SA1115 should be handled first
+                return;
+            }
+
+            for (int i = 2; i < parameters.Count; ++i)
+            {
+                ParameterSyntax currentParameter = parameters[i];
+                int currentLine = currentParameter.GetLine();
+
+                if (lineCondition(previousLine, currentLine))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, currentParameter.GetLocation()));
+                    return;
+                }
+
+                previousLine = currentLine;
+            }
+        }
+
+        private static void Analyze(SyntaxNodeAnalysisContext context, SeparatedSyntaxList<ArgumentSyntax> arguments)
+        {
+            ArgumentSyntax previousParameter = arguments[1];
+            int firstParameterLine = arguments[0].GetLine();
+            int previousLine = previousParameter.GetLine();
+            Func<int, int, bool> lineCondition;
+
+            if (firstParameterLine == previousLine)
+            {
+                // arguments must be on same line
+                lineCondition = (param1Line, param2Line) => param1Line != param2Line;
+            }
+            else if (previousLine - firstParameterLine == 1)
+            {
+                // each argument must be on its own line
+                lineCondition = (param1Line, param2Line) => param1Line == param2Line;
+            }
+            else
+            {
+                // SA1115 should be handled first
+                return;
+            }
+
+            for (int i = 2; i < arguments.Count; ++i)
+            {
+                ArgumentSyntax currentParameter = arguments[i];
+                int currentLine = currentParameter.GetLine();
+
+                if (lineCondition(previousLine, currentLine))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, currentParameter.GetLocation()));
+                    return;
+                }
+
+                previousLine = currentLine;
+            }
         }
```

Comment:
:thought_balloon:  There are quite a few places where almost the same code is used to check if an argument / parameter list has the correct layout. It would probably be possible to combine them in a single method with the signature below:

```c#
private static void Analyze(SyntaxNodeAnalysisContext context, IReadOnlyList<SyntaxNode> parameters)
```